File: PushLanguageInterpreter/PushLanguageInterpreter/Tokenizer.cs
using System.Text;

namespace PushLanguageInterpreter;

public class Tokenizer
{
    private int _position;

    public IReadOnlyList<Token> Tokenize(string program)
    {
        var instructions = new List<Token>();
        while (_position < program.Length)
        {
            if (IsToken(program, _position))
            {
                var word = new StringBuilder();
                while (IsToken(program, _position))
                {
                    word.Append(program[_position]);
                    _position++;
                }

                instructions.Add(new Token(word.ToString()));
            }
            else if (program[_position] == '(' || program[_position] == ')')
            {
                instructions.Add(new Token(program[_position].ToString()));
                _position++;
            }
            else if (char.IsWhiteSpace(program[_position]))
            {
                _position++;
            }
            else
            {
                throw new Exception("Unexpected character at position " + _position + ": " + program[_position]);
            }

        }

        return instructions;
    }

    private static bool IsToken(string program, int position) => char.IsLetter(program[position]) || program[position] == '_';
}

File: genetic-programming-scala-sandbox/untitled/src/main/java/com/yaskovdev/gp/push/Tokenizer.java
package com.yaskovdev.gp.push;

import java.util.ArrayList;
import java.util.List;

public class Tokenizer {

    public List<Token> tokenize(final String program) {
        int position = 0;
        final List<Token> tokens = new ArrayList<>();
        while (position < program.length()) {
            if (isToken(program.charAt(position))) {
                final StringBuilder word = new StringBuilder();
                while (isToken(program.charAt(position))) {
                    word.append(program.charAt(position));
                    position++;
                }
                tokens.add(new Token(word.toString()));
            } else if (program.charAt(position) == '(' || program.charAt(position) == ')') {
                tokens.add(new Token(Character.toString(program.charAt(position))));
                position++;
            } else if (Character.isWhitespace(program.charAt(position))) {
                position++;
            } else {
                throw new RuntimeException("Unexpected character at position " + position + ": " + program.charAt(position));
            }
        }
        return tokens;
    }

    private static boolean isToken(char character) {
        return Character.isLetterOrDigit(character) || character == '.' || character == '_' || character == '<';
    }
}
